File: backups/backup_20250903_042619/frontend/lib/security/security-headers.js
/**
 * Security Headers Configuration
 * Comprehensive security headers for production deployment
 */

const securityHeaders = [
  {
    key: 'X-DNS-Prefetch-Control',
    value: 'on',
  },
  {
    key: 'Strict-Transport-Security',
    value: 'max-age=63072000; includeSubDomains; preload',
  },
  {
    key: 'X-Frame-Options',
    value: 'DENY',
  },
  {
    key: 'X-Content-Type-Options',
    value: 'nosniff',
  },
  {
    key: 'Referrer-Policy',
    value: 'strict-origin-when-cross-origin',
  },
  {
    key: 'Permissions-Policy',
    value:
      'camera=(), microphone=(), geolocation=(), payment=(), usb=(), magnetometer=(), gyroscope=()',
  },
  {
    key: 'X-XSS-Protection',
    value: '1; mode=block',
  },
  {
    key: 'Content-Security-Policy',
    value: [
      "default-src 'self'",
      "script-src 'self' 'unsafe-inline' 'unsafe-eval' https://js.stripe.com https://www.googletagmanager.com https://www.google-analytics.com",
      "style-src 'self' 'unsafe-inline' https://fonts.googleapis.com",
      "img-src 'self' data: https: http:",
      "font-src 'self' https://fonts.gstatic.com",
      "connect-src 'self' https://api.ruleiq.com https://api-staging.ruleiq.com https://sentry.io https://www.google-analytics.com",
      "frame-src 'self' https://js.stripe.com",
      "object-src 'none'",
      "base-uri 'self'",
      "form-action 'self'",
    ].join('; '),
  },
];

// API-specific security headers
const apiSecurityHeaders = [
  ...securityHeaders,
  {
    key: 'Cache-Control',
    value: 'no-cache, no-store, must-revalidate',
  },
  {
    key: 'Pragma',
    value: 'no-cache',
  },
  {
    key: 'Expires',
    value: '0',
  },
  {
    key: 'X-Content-Type-Options',
    value: 'nosniff',
  },
  {
    key: 'X-RateLimit-Limit',
    value: '100',
  },
  {
    key: 'X-RateLimit-Remaining',
    value: '99', // This would be dynamic in real implementation
  },
];

// Development vs production headers
const getSecurityHeaders = (environment = 'production') => {
  if (environment === 'development') {
    return [
      ...securityHeaders.filter((h) => h.key !== 'Content-Security-Policy'),
      {
        key: 'Content-Security-Policy',
        value:
          "default-src * 'unsafe-inline' 'unsafe-eval'; script-src * 'unsafe-inline' 'unsafe-eval'; style-src * 'unsafe-inline';",
      },
    ];
  }

  return securityHeaders;
};

// CSP nonce generation for inline scripts
const generateCSPNonce = () => {
  return Buffer.from(crypto.randomBytes(16)).toString('base64');
};

// Security middleware for API routes
const securityMiddleware = (req, res, next) => {
  const headers = req.path.startsWith('/api') ? apiSecurityHeaders : getSecurityHeaders();

  headers.forEach((header) => {
    res.setHeader(header.key, header.value);
  });

  next();
};

// Security check utility
const checkSecurityHeaders = async (url) => {
  try {
    const response = await fetch(url, { method: 'HEAD' });
    const headers = response.headers;

    const checks = {
      'X-Frame-Options': headers.get('X-Frame-Options') === 'DENY',
      'X-Content-Type-Options': headers.get('X-Content-Type-Options') === 'nosniff',
      'Strict-Transport-Security': headers
        .get('Strict-Transport-Security')
        ?.includes('max-age=63072000'),
      'Content-Security-Policy': !!headers.get('Content-Security-Policy'),
      'Referrer-Policy': headers.get('Referrer-Policy') === 'strict-origin-when-cross-origin',
      'Permissions-Policy': !!headers.get('Permissions-Policy'),
      'X-XSS-Protection': headers.get('X-XSS-Protection') === '1; mode=block',
    };

    return {
      url,
      checks,
      score: (Object.values(checks).filter(Boolean).length / Object.keys(checks).length) * 100,
      missing: Object.keys(checks).filter((key) => !checks[key]),
    };
  } catch (error) {
    return {
      url,
      error: error.message,
      score: 0,
    };
  }
};

// Security configuration for Next.js
const securityConfig = {
  async headers() {
    return [
      {
        source: '/(.*)',
        headers: getSecurityHeaders(),
      },
      {
        source: '/api/(.*)',
        headers: apiSecurityHeaders,
      },
    ];
  },
};

module.exports = {
  securityHeaders,
  apiSecurityHeaders,
  getSecurityHeaders,
  generateCSPNonce,
  securityMiddleware,
  checkSecurityHeaders,
  securityConfig,
};

File: api/middleware/security_headers.py
"""
Security Headers Middleware for ruleIQ API.
Adds security headers to all HTTP responses to enhance application security.
"""

from __future__ import annotations

from typing import Any, Callable

from fastapi import Request
from starlette.responses import Response

# Import settings with fallback
try:
    from config.settings import settings
    is_production = getattr(settings, 'is_production', False)
    force_https = getattr(settings, 'force_https', False)
    debug = getattr(settings, 'debug', False)
except Exception:
    # Fallback if settings cannot be imported
    is_production = False
    force_https = False
    debug = False


async def security_headers_middleware(request: Request, call_next: Callable) -> Any:
    """
    Add security headers to all responses.
    
    This middleware adds various security headers to protect against
    common web vulnerabilities like XSS, clickjacking, and content sniffing.
    
    Args:
        request: The incoming HTTP request
        call_next: The next middleware or route handler
        
    Returns:
        The response with added security headers
    """
    # Process the request
    response = await call_next(request)
    
    # Common security headers for all environments
    response.headers['X-Content-Type-Options'] = 'nosniff'
    response.headers['X-Frame-Options'] = 'DENY'
    response.headers['X-XSS-Protection'] = '1; mode=block'
    response.headers['Referrer-Policy'] = 'strict-origin-when-cross-origin'
    response.headers['Permissions-Policy'] = (
        'geolocation=(), microphone=(), camera=(), payment=(), usb=(), '
        'magnetometer=(), gyroscope=(), accelerometer=()'
    )
    
    # Cache control for security
    if request.url.path.startswith('/api/'):
        response.headers['Cache-Control'] = 'no-store, no-cache, must-revalidate, private'
        response.headers['Pragma'] = 'no-cache'
        response.headers['Expires'] = '0'
    
    # Environment-specific headers
    # HSTS header should only be set in production/HTTPS environments
    if is_production or force_https:
        # Check if the request is actually over HTTPS (direct or via proxy)
        is_https = (
            request.url.scheme == 'https' or 
            request.headers.get('x-forwarded-proto') == 'https' or
            request.headers.get('x-forwarded-ssl') == 'on'
        )
        
        if is_https:
            # Strict Transport Security for HTTPS
            response.headers['Strict-Transport-Security'] = (
                'max-age=31536000; includeSubDomains; preload'
            )
        
        # Stricter CSP in production
        response.headers['Content-Security-Policy'] = (
            "default-src 'self'; "
            "script-src 'self' 'unsafe-inline' https://cdn.jsdelivr.net; "
            "style-src 'self' 'unsafe-inline' https://fonts.googleapis.com; "
            "font-src 'self' https://fonts.gstatic.com; "
            "img-src 'self' data: https:; "
            "connect-src 'self' https://api.ruleiq.com wss://api.ruleiq.com; "
            "frame-ancestors 'none'; "
            "base-uri 'self'; "
            "form-action 'self'"
        )
    else:
        # More permissive CSP for development
        response.headers['Content-Security-Policy'] = (
            "default-src 'self' 'unsafe-inline' 'unsafe-eval' http: https: ws: wss: data:; "
            "frame-ancestors 'none'"
        )
    
    # Cloud Run specific headers
    if request.headers.get('x-cloud-trace-context'):
        # Preserve Cloud Run trace context
        trace_context = request.headers.get('x-cloud-trace-context')
        if trace_context:
            response.headers['x-cloud-trace-context'] = trace_context
    
    # Remove potentially sensitive headers
    headers_to_remove = ['Server', 'X-Powered-By', 'X-AspNet-Version']
    for header in headers_to_remove:
        if header in response.headers:
            del response.headers[header]
    
    # Add custom security header for API version
    response.headers['X-API-Version'] = '1.0.0'
    
    # CORS preflight handling
    if request.method == 'OPTIONS':
        response.headers['Access-Control-Max-Age'] = '86400'  # 24 hours
    
    return response